packages/@react-spectrum/dialog/src/DialogContainer.tsx

```diff
@@ -0,0 +1,86 @@
+import {DialogContainerContext} from './DialogContainerContext';
+import React, {Fragment, ReactElement, useContext, useEffect, useRef, useState} from 'react';
+import {OverlayTriggerState, useOverlayTriggerState} from '@react-stately/overlays';
+import {SpectrumDialogClose, SpectrumDialogProps} from '@react-types/dialog';
+import {unwrapDOMRef, useMediaQuery} from '@react-spectrum/utils';
+import {Modal, Popover, Tray} from '@react-spectrum/overlays';
+import {DOMRefValue} from '@react-types/shared';
+import {useOverlayPosition, useOverlayTrigger} from '@react-aria/overlays';
+import {PressResponder} from '@react-aria/interactions';
+import {DialogContext} from './context';
+import {DialogTriggerBase} from './DialogTrigger';
+import {Dialog} from './Dialog';
+
+
+export function DialogContainer(props): ReactElement {
+  let {
+    children,
+    content,
+    type = 'modal',
+    mobileType = type === 'popover' ? 'modal' : type,
+    isDismissable
+  } = props;
+  // NOTE: content needs to be saved as function in order to save in useState (since it's a function as child)
+  let [overlayContent, setOverlayContent] = useState(() => content);
+  let [overlayProps, setOverlayProps] = useState(props);
+
+  // On small devices, show a modal or tray instead of a popover.
+  // TODO: DNA variable?
+  let isMobile = useMediaQuery('(max-width: 700px)');
+  if (isMobile) {
+    // handle cases where desktop popovers need a close button for the mobile modal view
+    if (type !== 'modal' && mobileType === 'modal') {
+      isDismissable = true;
+    }
+
+    type = mobileType;
+  }
+
+  let state = useOverlayTriggerState(overlayProps);
+
+  // todo just pass setOverlayContent via context value
+  let addContent = content => setOverlayContent(() => content);
+
+  let context = {
+    type,
+    onClose: state.close,
+    isDismissable
+  };
+  return (
+    <DialogContainerContext.Provider value={{isDismissable, state, overlayContent, addContent, setOverlayProps, type}}>
+      {children}
+      <DialogContext.Provider value={context}>
+        <DialogContainerOverlay />
```

Comment:
I think this might create an issue for nested dialogs if the overlay is inside the DialogContainerContext. The nested dialog trigger would consume from the context, see that a container is already there, and then reuse it. That would replace the outer dialog with the inner one rather than having both open at the same time.

Maybe you could only wrap `children` in the context rather than the overlay as well? Seems like the `DialogContainerOverlay` could be inlined into DialogContainer so it doesn't need the context. You could then use a fragment to put the context wrapping children next to the overlay rather than wrapping it.